pkg/retriever/cidretrieval.go

```diff
@@ -0,0 +1,502 @@
+	// OnErrorQueryingRetrievalCandidate may be called up to once per retrieval candidate
+	OnErrorQueryingRetrievalCandidate(candidate RetrievalCandidate, err error)
+	// OnErrorRetrievingFromCandidate may be called up to once per retrieval candidate
+	OnErrorRetrievingFromCandidate(candidate RetrievalCandidate, err error)
+	// OnRetrievalQueryForCandidate may be called up to once per retrieval candidate
+	OnRetrievalQueryForCandidate(candidate RetrievalCandidate, queryResponse *retrievalmarket.QueryResponse)
+	// OnFilteredRetrievalQueryForCandidate may be called up to once per retrieval candidate
+	OnFilteredRetrievalQueryForCandidate(candidate RetrievalCandidate, queryResponse *retrievalmarket.QueryResponse)
+	// OnRetrievingFromCandidate may be called up to once per retrieval candidate
+	OnRetrievingFromCandidate(candidate RetrievalCandidate)
+}
+
+type CidRetrieval interface {
+	RetrieveCid(context.Context, cid.Cid) (*RetrievalStats, error)
+}
+
+type runResult struct {
+	RetrievalResult *RetrievalStats
+	QueryError      error
+	RetrievalError  error
+}
+
+// queryCompare compares two QueryResponses and returns true if the first is
+// preferable to the second. This is used for the PriorityWaitQueue that will
+// prioritise execution of retrievals if two queries are available to compare
+// at the same time.
+var queryCompare prioritywaitqueue.ComparePriority[*retrievalmarket.QueryResponse] = func(a, b *retrievalmarket.QueryResponse) bool {
+	// Always prefer unsealed to sealed, no matter what
+	if a.UnsealPrice.IsZero() && !b.UnsealPrice.IsZero() {
+		return true
+	}
+
+	// Select lower price, or continue if equal
+	aTotalCost := totalCost(a)
+	bTotalCost := totalCost(b)
+	if !aTotalCost.Equals(bTotalCost) {
+		return aTotalCost.LessThan(bTotalCost)
+	}
+
+	// Select smaller size, or continue if equal
+	if a.Size != b.Size {
+		return a.Size < b.Size
+	}
+
+	return false
+}
+
+var _ CidRetrieval = (*cidRetrieval)(nil)
+
+type cidRetrieval struct {
+	Ctx                         context.Context
+	IndexEndpoint               Endpoint
+	Client                      RetrievalClient
+	Instrumentation             Instrumentation
+	GetStorageProviderTimeout   GetStorageProviderTimeout
+	IsAcceptableStorageProvider IsAcceptableStorageProvider
+	IsAcceptableQueryResponse   IsAcceptableQueryResponse
+
+	WaitGroup sync.WaitGroup
+}
+
+// retrievalState is used on a per-retrieval basis so we can handle multiple
+// retrievals on a single cidRetrieval instance
+type retrievalState struct {
+	Cid                 cid.Cid
+	WaitQueue           prioritywaitqueue.PriorityWaitQueue[*retrievalmarket.QueryResponse]
+	ResultChan          chan runResult
+	FinishChan          chan struct{}
+	ActiveQueryCancel   []*func()
+	ActiveQueryCancelLk sync.Mutex
+}
+
+// NewCidRetrieval creates a new CidRetrieval. A CidRetrieval instance may be
+// used for multiple retrievals via the RetrieveCid call.
+func NewCidRetrieval(
```

Comment:
Public function returning private type is a bit of an anti-pattern. Any reason it's a private struct when all it's members are public?

Reply:
ok, so should I switch it around, so the type is public but the fields are private? I wanted it all to be private but the "public" methods on it were just a way of making them stand out in the code, a weak attempt at making the code less ugly; but I don't think it achieved much.